include/rect.h

```diff
@@ -21,41 +21,186 @@
 #ifndef DOSBOX_RECTANGLE_H
 #define DOSBOX_RECTANGLE_H
 
-#include "math_utils.h"
+#include <algorithm>
+#include <cassert>
 
 namespace DosBox {
 
-// Struct to represent rectangles.
+// Struct to represent rectangles and sizes.
+//
+// As a general practice, it should be encoded in the variable/argument name
+// what we're dealing with (e.g. `viewport_rect`, `desktop_size`).
+//
 struct Rect {
 	constexpr Rect() = default;
 
-	constexpr Rect(const float _w, const float _h)
-	        : x(0.0f),
-	          y(0.0f),
-	          w(_w),
-	          h(_h)
-	{}
+	constexpr Rect(const float x_pos, const float y_pos, const float width,
+	               const float height)
+	        : x(x_pos),
+	          y(y_pos),
+	          w(width),
+	          h(height)
+	{
+		assert(w >= 0.0f);
+		assert(h >= 0.0f);
+	}
```

Comment:
These asserts are good, but because it's a `struct`, the API encourages directly working with member variables. 

```
// some bit where we update frame size..
window_frame.w = (math based on some variables);
window_frame.h = (math based on some variables);
```

This on-the-fly adjustment is where we could accidentally violate the asserts (if something returned a negative, even if it's very small), leading to the polygon overlap / intersect / etc.. routines not being right.

One option is making them private and asserting on the only path that's allowed to change them.

Reply:
Good points, and seem good suggestions on the surface, but the more I think about, I'm coming to the conclusion I want this `Rectangle` class to be flexible and support generic use cases. That includes rectangles with zero dimensions, and possibly even negative dimensions. These are just useful in generic algorithms, same way an "empty list/container" is useful. A rectangle can be similarly "empty" if its dimensions are zero, and negative dimensions are valid for certain algorithms and/or for intermediate results.

I'm considering turning this into something similar to the Rectangle class from the Java standard library which allows and handles zero and negative dimensions. It's worthwhile to read the intro paragraph, but some key points are below:
https://docs.oracle.com/javase/8/docs/api/java/awt/Rectangle.html

> A Rectangle whose width or height is exactly zero has location along those axes with zero dimension, but is otherwise considered empty. The [isEmpty()](https://docs.oracle.com/javase/8/docs/api/java/awt/Rectangle.html#isEmpty--) method will return true for such a Rectangle.

> Methods which test if an empty Rectangle contains or intersects a point or rectangle will always return false if either dimension is zero.

> Methods which combine such a Rectangle with a point or rectangle will include the location of the Rectangle on that axis in the result as if the [add(Point)](https://docs.oracle.com/javase/8/docs/api/java/awt/Rectangle.html#add-java.awt.Point-) method were being called.

> A Rectangle whose width or height is negative [...] is also empty with respect to containment calculations

So I'm thinking of allowing everything, keeping the direct property access which is desirable, and then asserting a non-empty rectangle is the caller's responsibility if such restriction is desired (varies on a case by case basis):

```cpp
assert(rect.IsNonEmpty());
```

That moves the assert up the chain to its correct place, and we keep our flexible rectangle implementation.

Reply:
All sounds good; and like that approach.
Just having a generic rectangle struct is very handy - even if you only use as a collection instead of a functional object.

Yeah - for the functions, like Intersect, that's where you could add the `assertm`'s and mention the intersection algorithm only handles positive WxH values.